File: packages/observability-py/domio_obs/internal/metrics.py
"""
Metrics API — minimal OTLP-flavored counters, histograms, and gauges.
"""

from __future__ import annotations

import re
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple

from domio_obs.internal.exporter import OtlpHttpExporter
from domio_obs.internal.redaction import redact_value
from domio_obs.internal.resource import ResourceAttributes

SAFE_NAME_RE = re.compile(r"^[a-zA-Z][a-zA-Z0-9_.]*$")

DEFAULT_BUCKETS_MS: Tuple[float, ...] = (
    1, 2, 5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000, 30000,
)


@dataclass
class MetricOptions:
    description: str = ""
    unit: str = ""


class Counter:
    def __init__(self, name: str, meter: "Meter", options: Optional[MetricOptions] = None) -> None:
        self.name = name
        self._meter = meter
        self._opts = options or MetricOptions()

    def add(self, value: float, attributes: Optional[Dict[str, str]] = None) -> None:
        self._meter._add_counter(self.name, value, attributes, monotonic=True)


class UpDownCounter:
    def __init__(self, name: str, meter: "Meter", options: Optional[MetricOptions] = None) -> None:
        self.name = name
        self._meter = meter
        self._opts = options or MetricOptions()

    def add(self, value: float, attributes: Optional[Dict[str, str]] = None) -> None:
        self._meter._add_counter(self.name, value, attributes, monotonic=False)


class Histogram:
    def __init__(self, name: str, meter: "Meter", options: Optional[MetricOptions] = None) -> None:
        self.name = name
        self._meter = meter
        self._opts = options or MetricOptions()

    def record(self, value: float, attributes: Optional[Dict[str, str]] = None) -> None:
        self._meter._record_histogram(self.name, value, attributes)


def _attr_key(attrs: Optional[Dict[str, str]]) -> str:
    if not attrs:
        return ""
    keys = sorted(attrs.keys())
    return ",".join(f"{k}={attrs[k]}" for k in keys)


class Meter:
    def __init__(self, resource: ResourceAttributes, exporter: Optional[OtlpHttpExporter]) -> None:
        self.resource = resource
        self.exporter = exporter
        self._counters: Dict[str, Dict[str, float]] = {}
        self._counter_opts: Dict[str, MetricOptions] = {}
        self._counter_meta: Dict[str, bool] = {}
        self._hists: Dict[str, Dict[str, List[float]]] = {}
        self._hist_opts: Dict[str, MetricOptions] = {}
        self._flushing = False

    def create_counter(self, name: str, options: Optional[MetricOptions] = None) -> Counter:
        if not SAFE_NAME_RE.match(name):
            raise ValueError(f"invalid metric name: {name!r}")
        self._counters.setdefault(name, {})
        self._counter_opts[name] = options or MetricOptions()
        self._counter_meta[name] = True
        return Counter(name, self, options)

    def create_up_down_counter(self, name: str, options: Optional[MetricOptions] = None) -> UpDownCounter:
        if not SAFE_NAME_RE.match(name):
            raise ValueError(f"invalid metric name: {name!r}")
        self._counters.setdefault(name, {})
        self._counter_opts[name] = options or MetricOptions()
        self._counter_meta[name] = False
        return UpDownCounter(name, self, options)

    def create_histogram(self, name: str, options: Optional[MetricOptions] = None) -> Histogram:
        if not SAFE_NAME_RE.match(name):
            raise ValueError(f"invalid metric name: {name!r}")
        self._hists.setdefault(name, {})
        self._hist_opts[name] = options or MetricOptions()
        return Histogram(name, self, options)

    def _add_counter(self, name: str, value: float, attributes: Optional[Dict[str, str]], monotonic: bool) -> None:
        self._counters.setdefault(name, {})
        k = _attr_key(attributes)
        self._counters[name][k] = self._counters[name].get(k, 0.0) + value

    def _record_histogram(self, name: str, value: float, attributes: Optional[Dict[str, str]]) -> None:
        self._hists.setdefault(name, {})
        k = _attr_key(attributes)
        slot = self._hists[name].get(k)
        if slot is None:
            slot = [0.0] * len(DEFAULT_BUCKETS_MS) + [0.0]  # trailing overflow bucket
            self._hists[name][k] = slot
        for i, bound in enumerate(DEFAULT_BUCKETS_MS):
            if value <= bound:
                slot[i] += 1
        slot[-1] += 1  # total observations always incremented

    def flush(self) -> None:
        if self.exporter is None:
            return
        if self._flushing:
            return
        self._flushing = True
        try:
            counters = self._counters
            hists = self._hists
            self._counters = {}
            self._hists = {}
            metric_payloads: List[Dict[str, Any]] = []

            for name, by_attrs in counters.items():
                opts = self._counter_opts.get(name, MetricOptions())
                monotonic = self._counter_meta.get(name, True)
                for k, v in by_attrs.items():
                    metric_payloads.append(
                        {
                            "name": name,
                            "description": opts.description,
                            "unit": opts.unit,
                            "sum": v,
                            "aggregationTemporality": 2,
                            "isMonotonic": monotonic,
                            "attributes": _string_attrs_to_otlp(_parse_attr_key(k)),
                        }
                    )

            for name, by_attrs in hists.items():
                opts = self._hist_opts.get(name, MetricOptions())
                for k, slot in by_attrs.items():
                    # slot layout: [bucket0, bucket1, ..., bucketN-1, total]
                    # bucketCounts in OTLP is cumulative — each entry is
                    # the count of observations whose value is <= the
                    # bound at the same index. The trailing overflow
                    # bucket equals the total count of observations.
                    bounds = list(DEFAULT_BUCKETS_MS)
                    counts = list(slot[:-1]) + [float(slot[-1])]
                    total = float(slot[-1])
                    metric_payloads.append(
                        {
                            "name": name,
                            "description": opts.description,
                            "unit": opts.unit,
                            "count": total,
                            "sum": 0.0,
                            "aggregationTemporality": 2,
                            "isMonotonic": True,
                            "bucketBounds": bounds,
                            "bucketCounts": counts,
                            "attributes": _string_attrs_to_otlp(_parse_attr_key(k)),
                        }
                    )

            if not metric_payloads:
                return

            payload = {
                "resourceMetrics": [
                    {
                        "resource": self.resource.to_otlp(),
                        "scopeMetrics": [
                            {
                                "scope": {"name": "@domio/observability-py"},
                                "metrics": metric_payloads,
                            }
                        ],
                    }
                ]
            }
            self.exporter.export_json("metrics", payload)
        finally:
            self._flushing = False

    def shutdown(self) -> None:
        self.flush()
        if self.exporter is not None:
            self.exporter.shutdown()


def _string_attrs_to_otlp(attrs: Dict[str, str]) -> List[Dict[str, Any]]:
    return [{"key": k, "value": {"stringValue": v}} for k, v in attrs.items()]


def _parse_attr_key(k: str) -> Dict[str, str]:
    out: Dict[str, str] = {}
    if not k:
        return out
    for pair in k.split(","):
        idx = pair.find("=")
        if idx <= 0:
            continue
        out[pair[:idx]] = pair[idx + 1:]
    return out

File: packages/observability/src/metrics.ts
/**
 * Metrics API — minimal OTLP-flavored counters, histograms, and gauges.
 *
 * Public API:
 *   - meter.createCounter(name, opts?)
 *   - meter.createHistogram(name, opts?)
 *   - meter.createUpDownCounter(name, opts?)
 *   - meter.flush() / meter.shutdown()
 *
 * Counters and UpDownCounters accumulate values; Histograms keep a
 * fixed bucket layout (exponential, base 2). On flush, every registered
 * instrument is serialized into a single OTLP `ExportMetricsServiceRequest`
 * and posted to the exporter.
 */

import type { ResourceAttributes } from './resource.js';
import type { OtlpHttpExporter } from './exporters/otlp-http.js';
import { resourceToOtlp } from './exporters/otlp-http.js';
import { getRedactor } from './redaction.js';

export interface Counter {
  type: 'counter';
  name: string;
  description?: string;
  unit?: string;
  add(value: number, attrs?: Record<string, string>): void;
}

export interface UpDownCounter {
  type: 'up_down_counter';
  name: string;
  description?: string;
  unit?: string;
  add(value: number, attrs?: Record<string, string>): void;
}

export interface Histogram {
  type: 'histogram';
  name: string;
  description?: string;
  unit?: string;
  record(value: number, attrs?: Record<string, string>): void;
}

export type Instrument = Counter | UpDownCounter | Histogram;

export interface Meter {
  createCounter(name: string, opts?: { description?: string; unit?: string }): Counter;
  createHistogram(name: string, opts?: { description?: string; unit?: string }): Histogram;
  createUpDownCounter(name: string, opts?: { description?: string; unit?: string }): UpDownCounter;
  flush(): Promise<void>;
  shutdown(): Promise<void>;
  readonly resource: ResourceAttributes;
  readonly exporter: OtlpHttpExporter | null;
}

export interface MeterConfig {
  resource: ResourceAttributes;
  exporter: OtlpHttpExporter | null;
}

const SAFE_NAME_RE = /^[a-zA-Z][a-zA-Z0-9_.]*$/;

/**
 * Default histogram buckets. They cover web request latencies from
 * ~1ms to ~16s with reasonable resolution in the SLO-relevant zone
 * (50ms..2s).
 */
export const DEFAULT_BUCKETS_MS: readonly number[] = Object.freeze([
  1, 2, 5, 10, 25, 50, 100, 250, 500, 1_000, 2_500, 5_000, 10_000, 30_000,
]);

interface CounterState {
  type: 'counter' | 'up_down_counter';
  name: string;
  description?: string;
  unit?: string;
  byAttrs: Map<string, number>;
}

interface HistogramBucket {
  count: number;
  sum: number;
  buckets: number[];
}

interface HistogramState {
  type: 'histogram';
  name: string;
  description?: string;
  unit?: string;
  byAttrs: Map<string, HistogramBucket>;
  bucketBounds: readonly number[];
  buckets: number[];
}

export function createMeter(cfg: MeterConfig): Meter {
  const counters = new Map<string, CounterState>();
  const histograms = new Map<string, HistogramState>();
  let flushing = false;

  function getOrCreateCounter(
    name: string,
    type: 'counter' | 'up_down_counter',
    description?: string,
    unit?: string,
  ): CounterState {
    if (!SAFE_NAME_RE.test(name)) throw new Error(`invalid metric name: ${name}`);
    const existing = counters.get(name);
    if (existing) return existing;
    const state: CounterState = { type, name, byAttrs: new Map() };
    if (description !== undefined) state.description = description;
    if (unit !== undefined) state.unit = unit;
    counters.set(name, state);
    return state;
  }

  function getOrCreateHistogram(name: string, description?: string, unit?: string): HistogramState {
    if (!SAFE_NAME_RE.test(name)) throw new Error(`invalid metric name: ${name}`);
    const existing = histograms.get(name);
    if (existing) return existing;
    const buckets = DEFAULT_BUCKETS_MS.map(() => 0);
    const state: HistogramState = {
      type: 'histogram',
      name,
      bucketBounds: DEFAULT_BUCKETS_MS,
      buckets,
      byAttrs: new Map(),
    };
    if (description !== undefined) state.description = description;
    if (unit !== undefined) state.unit = unit;
    histograms.set(name, state);
    return state;
  }

  function attrKey(attrs: Record<string, string> | undefined): string {
    if (!attrs) return '';
    const keys = Object.keys(attrs).sort();
    return keys.map((k) => `${k}=${attrs[k] ?? ''}`).join(',');
  }

  function decorateOpts(target: { description?: string; unit?: string }, opts?: { description?: string; unit?: string }) {
    if (opts?.description !== undefined) target.description = opts.description;
    if (opts?.unit !== undefined) target.unit = opts.unit;
  }

  function createCounter(name: string, opts?: { description?: string; unit?: string }): Counter {
    const state = getOrCreateCounter(name, 'counter', opts?.description, opts?.unit);
    const counter: Counter = {
      type: 'counter',
      name,
      add(value, attrs) {
        const k = attrKey(attrs);
        state.byAttrs.set(k, (state.byAttrs.get(k) ?? 0) + value);
      },
    };
    decorateOpts(counter, opts);
    return counter;
  }

  function createUpDownCounter(name: string, opts?: { description?: string; unit?: string }): UpDownCounter {
    const state = getOrCreateCounter(name, 'up_down_counter', opts?.description, opts?.unit);
    const udc: UpDownCounter = {
      type: 'up_down_counter',
      name,
      add(value, attrs) {
        const k = attrKey(attrs);
        state.byAttrs.set(k, (state.byAttrs.get(k) ?? 0) + value);
      },
    };
    decorateOpts(udc, opts);
    return udc;
  }

  function createHistogram(name: string, opts?: { description?: string; unit?: string }): Histogram {
    const state = getOrCreateHistogram(name, opts?.description, opts?.unit);
    const h: Histogram = {
      type: 'histogram',
      name,
      record(value, attrs) {
        const k = attrKey(attrs);
        let bucket = state.byAttrs.get(k);
        if (!bucket) {
          bucket = { count: 0, sum: 0, buckets: state.bucketBounds.map(() => 0) };
          state.byAttrs.set(k, bucket);
        }
        bucket.count += 1;
        bucket.sum += value;
        for (let i = 0; i < state.bucketBounds.length; i++) {
          const bound = state.bucketBounds[i];
          if (bound === undefined) continue;
          if (value <= bound) {
            const b = bucket.buckets[i];
            if (b !== undefined) bucket.buckets[i] = b + 1;
          }
        }
      },
    };
    decorateOpts(h, opts);
    return h;
  }

  return {
    resource: cfg.resource,
    exporter: cfg.exporter,
    createCounter,
    createUpDownCounter,
    createHistogram,
    async flush() {
      if (!cfg.exporter) return;
      if (flushing) return;
      flushing = true;
      try {
        const redactor = getRedactor();
        const dataPoints: unknown[] = [];

        for (const state of counters.values()) {
          for (const [k, v] of state.byAttrs) {
            const attrs = parseAttrKey(k);
            const dp: Record<string, unknown> = {
              name: state.name,
              sum: v,
              aggregationTemporality: 2,
              isMonotonic: state.type === 'counter',
              attributes: attrToOtlp(redactor.redactValue(attrs)),
            };
            if (state.description !== undefined) dp['description'] = state.description;
            if (state.unit !== undefined) dp['unit'] = state.unit;
            dataPoints.push(dp);
            state.byAttrs.set(k, 0); // reset after delta
          }
        }

        for (const state of histograms.values()) {
          for (const [k, v] of state.byAttrs) {
            const attrs = parseAttrKey(k);
            // bucketCounts is cumulative (per the OTLP spec). The
            // trailing overflow bucket equals `count` (the total number
            // of observations); all observations fall into the overflow
            // bucket by definition, since the bucket array has no upper
            // bound on it.
            const dp: Record<string, unknown> = {
              name: state.name,
              count: v.count,
              sum: v.sum,
              aggregationTemporality: 2,
              bucketBounds: [...state.bucketBounds],
              bucketCounts: [...v.buckets, v.count],
              attributes: attrToOtlp(redactor.redactValue(attrs)),
            };
            if (state.description !== undefined) dp['description'] = state.description;
            if (state.unit !== undefined) dp['unit'] = state.unit;
            dataPoints.push(dp);
            state.byAttrs.set(k, { count: 0, sum: 0, buckets: state.bucketBounds.map(() => 0) });
          }
        }

        if (dataPoints.length === 0) return;

        const payload = {
          resourceMetrics: [
            {
              resource: resourceToOtlp(cfg.resource),
              scopeMetrics: [
                {
                  scope: { name: '@domio/observability' },
                  metrics: dataPoints,
                },
              ],
            },
          ],
        };
        await cfg.exporter.exportJson('metrics', payload);
      } finally {
        flushing = false;
      }
    },
    async shutdown() {
      await this.flush();
      if (cfg.exporter) await cfg.exporter.shutdown();
    },
  };
}

function attrToOtlp(attrs: Record<string, string>): Array<{ key: string; value: { stringValue: string } }> {
  return Object.entries(attrs).map(([key, value]) => ({ key, value: { stringValue: value } }));
}

function parseAttrKey(k: string): Record<string, string> {
  if (k.length === 0) return {};
  const out: Record<string, string> = {};
  for (const pair of k.split(',')) {
    const idx = pair.indexOf('=');
    if (idx <= 0) continue;
    out[pair.slice(0, idx)] = pair.slice(idx + 1);
  }
  return out;
}
